File: Moobeez/Classes/Moobeez/Models/TimelineItem.m
//
//  TimelineItem.m
//  Moobeez
//
//  Created by Radu Banea on 12/08/14.
//  Copyright (c) 2014 Goggzy. All rights reserved.
//

#import "TimelineItem.h"
#import "Moobeez.h"

@interface TimelineItem ()

@end

@implementation TimelineItem

- (id)initWithDatabaseDictionary:(NSDictionary*)databaseDictionary {
    
    self = [super initWithDatabaseDictionary:databaseDictionary];
    
    if (self) {

        self.name = databaseDictionary[@"name"];
        self.backdropPath = databaseDictionary[@"backdropPath"];
        if (databaseDictionary[@"date"]) {
            if (databaseDictionary[@"episodeNumber"]) {
                self.date = [[NSDate dateWithTimeIntervalSince1970:[databaseDictionary[@"date"] doubleValue]] teebeeDisplayDate];
            }
            else {
                self.date = [NSDate dateWithTimeIntervalSince1970:[databaseDictionary[@"date"] doubleValue]];
            }
        }
        
        if (databaseDictionary[@"seasonNumber"]) {
            self.season = [databaseDictionary[@"seasonNumber"] integerValue];
        }
        else {
            self.season = -1;
        }
        
        if (databaseDictionary[@"episodeNumber"]) {
            self.episode = [databaseDictionary[@"episodeNumber"] integerValue];
        }
        
        self.tmdbId = [databaseDictionary[@"tmdbId"] integerValue];
        
        if (databaseDictionary[@"rating"]) {
            self.rating = [databaseDictionary[@"rating"] floatValue];
        }
        else {
            self.rating = -1;
        }

    }

    return self;
}

#pragma mark - Comparison selectors

- (NSComparisonResult)compareByDate:(TimelineItem*)timelineItem {
    return [self.date compare:timelineItem.date];
}

- (NSComparisonResult)compareDescByDate:(TimelineItem*)timelineItem {
    return [timelineItem.date compare:self.date];
}

- (NSComparisonResult)compareById:(TimelineItem*)timelineItem {
    return [[NSNumber numberWithInteger:timelineItem.id] compare:[NSNumber numberWithInteger:self.id]];
}

- (BOOL)isEqual:(id)object {
    if ([object isKindOfClass:[self class]]) {
        if (self.id == ((TimelineItem*) object).id) {
            return YES;
        }
    }
    
    return NO;
}

- (BOOL)isMovie {
    return self.season == -1;
}

@end

File: Moobeez/Classes/Models/TimelineItem.swift
//
//  TimelineItem.swift
//  Moobeez
//
//  Created by Radu Banea on 24/02/2018.
//  Copyright © 2018 Gogo's Lab. All rights reserved.
//

import UIKit

class TimelineItem: NSObject {

    convenience init(moobee:Moobee) {
        self.init()
        self.moobee = moobee
        
        if moobee.type == .seen {
            date = moobee.date
        }
        else {
            date = moobee.releaseDate
        }
        
        date = Calendar.current.startOfDay(for: date!)
        
        dateString = date?.string(withFormat: "YYYY-MM-DD")
    }
    
    convenience init(episode:Teebee.Episode) {
        self.init()
        self.teebeeEpisode = episode
        
        date = teebeeEpisode?.releaseDate
        
        date = Calendar.current.startOfDay(for: date!)
        
        dateString = date?.string(withFormat: "YYYY-MM-DD")
    }
    
    var moobee:Moobee?
    
    var teebeeEpisode:Teebee.Episode?
    
    var name:String? {
        get {
            if moobee != nil {
                return moobee!.name
            }
            
            if teebeeEpisode != nil {
                return teebeeEpisode!.season!.teebee!.name
            }
            
            return nil
        }
    }
    
    var subtitle:String? {
        get {
            if teebeeEpisode != nil {
                return "Season \(teebeeEpisode!.season!.number) Episode \(teebeeEpisode!.number)"
            }
            
            return nil
        }
    }

    var date:Date?
    var dateString:String?
    
    var backdropPath:String? {
        get {
            if moobee != nil {
                return moobee!.backdropPath
            }
            
            if teebeeEpisode != nil {
                return teebeeEpisode!.season!.teebee!.backdropPath
            }
            
            return nil
        }
    }
    
    var watched:Bool? {
        get {
            if moobee != nil {
                return moobee!.type == .seen
            }
            
            if teebeeEpisode != nil {
                return teebeeEpisode!.watched
            }
            
            return nil
        }
    }
    
    var rating:Float {
        if moobee != nil {
            if moobee!.type == .seen {
                return moobee!.rating
            }
        }
        
        return -1
    }
    
}
